_posts/2015-12-30-Python-Join-Tutorial1.md

```diff
@@ -0,0 +1,302 @@
+---
+title: Real World Hadoop - Real World Hadoop Guide for Python 
+layout: post
+description: Data processing pipelines with Spark
+topic: engineering
+author: matthew_rathbone
+coauthor:
+  name: Elena Akhmatova
+  link: https://ru.linkedin.com/pub/elena-akhmatova/3/877/266
+categories:
+- hadoop
+tags:
+- python
+- hadoop
+- hadoop-streaming
+---
+
+> This article is part of [my guide to map reduce frameworks][4] in which I implement a solution to a real-world problem in each of the most popular Hadoop frameworks.  
+>  
+> One of the articles in the guide [Hadoop Python MapReduce Tutorial for Beginners][13] has already introduced the reader to the basics of hadoop-streaming with Python. This is the next logical step in a quest to learn how to use Python in map reduce framework defined by Hadoop.  
+> 
+
+## The Problem
+
+Let me quickly restate the problem from [my original article][6].
+
+I have two datasets:
+
+1. User information (id, email, language, location)
+2. Transaction information (transaction-id, product-id, user-id, purchase-amount, item-description)
+
+Given these datasets, I want to find the number of unique locations in which each product has been sold. To do that, I need to join the two datasets together.
+
+Previously I have implemented this solution [in java][7], [with hive][8] and [with pig][9]. The java solution was ~500 lines of code, hive and pig were like ~20 lines tops.
+
+The [Python article][13] i mentioned earlier does not solve this problem. It has a focus on understanding the basics of working with Hadoop-streaming and Python. 
+
+## The Python Solution
+
+Again, this article is a follow up for my [earlier article][13] on [Python][1] that shows the basics of using Python for map reduce in Hadoop. Please refer to it to gather basic understanding on how to write mappers and reducers in Python for Hadoop. Here we focus on implementing join and count distinct by key in Python and will assume some preliminary understanding of hadoop-streaming. 
+
+## Demonstration Data
+
+As previously (and here we mean "versus" the solutions [in java][7], [with hive][8] and [with pig][9]) we use two datasets called `users` and `transactions`. 
+
+{% highlight bash %}
+users
+1	matthew@test.com	EN	US
+2	matthew@test2.com	EN	GB
+3	matthew@test3.com	FR	FR
+{% endhighlight %}
+
+and
+
+{% highlight bash %}
+transactions
+1	1	1	300	a jumper
+2	1	2	300	a jumper
+3	1	2	300	a jumper
+4	2	3	100	a rubber chicken
+5	1	3	300	a jumper
+{% endhighlight %}
+
+**This time though they will constitute the same input dataset!** Soon it will become clear why.
+
+{% highlight bash %}
+hdfs dfs -mkdir input
+
+hdfs dfs -put ./users.txt input
+hdfs dfs -put ./transactions.txt input
+{% endhighlight %}
+
+And for command line testing
```

Comment:
Hi Matthew, 

I have also started thinking about the Type Safe versus Fields Based article. I think it will unfold somehow like the following:

These article is a second article on Scalding and should be read together with the … article.  
There are several types of API for Scalding: Type Safe API, Fields based API, and Matrix API. The previous article solves the sample task of my guide to mapreduce frameworks using Fields based API. This time we solve the same problem using Type Safe API.

=== BEGIN
To think about the differences here we should remind ourselves that Scalding was build on Cascading. Cascading has a record model where each element in underlying distributed list/table is a table with some named fields. Hence Scalding inherited this from Cascading and used it allowing us to define the named fields, schema-like. 

Before (using Fileds Based API) we directly specified the schema of fields we expected:
val users = ( 'id, 'email, 'language, 'location)
val transactions = ( 'transaction_id, 'product_id, 'user_id, 'purchase_amount, 'item_description)

The problem here is the type inference at compile time. As Fields based API allows to name the fields but the type of the fields is not specified different latent problems might emerge during run time. 

Type Safe API is newer than Fields Based API and it addresses that issue.

=== END

But I think the code I have submitted today is somehow not exactly right, even though it compiles and tests do pass. Do you have any suggestions on what to read about the differences between the two APIs to understand better what is happening?

So far I have:
original articles:
https://github.com/twitter/scalding/wiki/Matrix-API-Reference
https://github.com/twitter/scalding/wiki/Fields-based-API-Reference
https://github.com/twitter/scalding/wiki/Type-safe-api-reference

Shortish but helpful articles:
http://scalding.readthedocs.org/en/readthedocs/gen/other/Comparison-to-Scrunch-and-Scoobi.html
https://www.quora.com/What-are-the-differences-between-Crunch-and-Cascading

Useless but existing articles:
http://no-kill-switch.ghost.io/scalding-functional-data-processing-using-scala-hadoop/
http://what-when-how.com/Tutorial/topic-527dsksibr/Enterprise-Data-Workflows-with-Cascading-79.html

Regards, 
Elena.